packages/model-viewer/src/three-components/ModelUtils.ts

```diff
@@ -36,10 +32,35 @@ export const cloneGltf = (gltf: GLTF): GLTF => {
    */
   const cloneAndPatchMaterial = (material: Material): Material => {
     const clone = material.clone();
+    // This allows us to patch three's materials, on top of patches already
+    // made, for instance GLTFLoader patches SpecularGlossiness materials.
+    // Unfortunately, three's program cache differentiates SpecGloss materials
+    // via onBeforeCompile.toString(), so these two functions do the same thing
+    // but look different in order to force a proper recompile.
+    const oldOnBeforeCompile = material.onBeforeCompile;
+    clone.onBeforeCompile = (material as any).isGLTFSpecularGlossinessMaterial ?
```

Comment:
Is it possible for both material types to use the the SpecularGlossiness path here?

Reply:
Alas no; I tried to describe it in the comment. If the toString() version of onBeforeCompile is the same for the two materials, the second one will not be compiled and will instead try to use the first, which does not work. Took me all morning to diagnose this bug...

Reply:
Yah, that's a really funky way to test for shader equivalence.

It would be nice if there was a more hygienic / extensible API for patching Three.js materials. Oh well.

Reply:
Indeed, though I'm just glad there's a way to patch Three.js materials at all. The onBeforeCompile method hasn't been around that long.